File: chapter4/max/1.go
package main

import "fmt"

func max(list []int) int {
	length := len(list)

	var maxNum int

	if length < 2 {
		return list[0]
	}

	if length == 2 {
		maxNum = list[1]
	} else {
		maxNum = max(list[1:length])
	}

	if list[0] < maxNum {
		return maxNum
	} else {
		return list[0]
	}
}

func main() {
	fmt.Println(max([]int{100, 1203, 0, -10, 20, 8, 145}))
}

File: chapter4/max/1.exs
defmodule Grok do
  # TODO: Не нравится мне как это тут устроено всё... 
  def max([]), do: nil
  def max([head | tail]) when length(tail) == 2 do
    tail_el = Enum.at(tail, 0)

    if head < tail_el, do: tail_el, else: head 
  end 
  def max([head | tail]) do
    max_num = max(tail)

    if head < max_num, do: max_num, else: head 
  end
end

IO.puts Grok.max([100, 1203, 0, -10, 20, 8, 145])